src/docker-compose

```diff
@@ -0,0 +1,318 @@
+        get_manifest_image_ids "$manifest" >> "${PERSISTENT_STORE}/new/image_ids"
+    done
+    if [ $rc -ne 0 ]; then
+        return $rc
+    fi
+
+    if test -d "${PERSISTENT_STORE}/current"; then
+        echo "saving existing composition as -previous"
+        rm -rf "${PERSISTENT_STORE}/previous" # should not be there, but just in case
+        mv -v "${PERSISTENT_STORE}/current" "${PERSISTENT_STORE}/previous"
+    else
+        echo "no previous composition found"
+    fi
+
+    # If the previous version is present, we need to make sure it's stopped.
+    # XXX: let docker compose figure this out?
+    if test -d "${PERSISTENT_STORE}/previous"; then
+        comp_stop "${PERSISTENT_STORE}/previous"
+    else
+        echo "previous composition not present; nothing to stop."
+    fi
+
+    comp_start "${PERSISTENT_STORE}/new" 2>&1
+}
+
+commit_artifact() {
+    mv -v "${PERSISTENT_STORE}/new" "${PERSISTENT_STORE}/current"
+    if [ -d "${PERSISTENT_STORE}/previous" ]; then
+        mv -v "${PERSISTENT_STORE}/previous" "${PERSISTENT_STORE}/cleanup"
+    fi
+}
+
+rollback_artifact() {
+    if ! test -d "${PERSISTENT_STORE}/new"; then
+        if test -d "${PERSISTENT_STORE}/current"; then
+            # This may seem weird (and it is!), but rollback can be requested even
+            # after a commit. In that case we rollback from the committed version to
+            # the one pushed for cleanup.
+            echo "Rolling back after a Commit"
+
+            # Let's set the stage for the rest of this function.
+            mv -v "${PERSISTENT_STORE}/current" "${PERSISTENT_STORE}/new"
+            if [ -d "${PERSISTENT_STORE}/cleanup" ]; then
+                mv -v "${PERSISTENT_STORE}/cleanup" "${PERSISTENT_STORE}/previous"
+            fi
+        else
+            echo "Nothing to rollback to!"
+            return 1
+        fi
+    fi
+
+    # Stopping may fail in case starting failed which we don't know.
+    comp_stop "${PERSISTENT_STORE}/new" || true
+    mv -v "${PERSISTENT_STORE}/new" "${PERSISTENT_STORE}/cleanup"
+
+    if [ -d "${PERSISTENT_STORE}/previous" ]; then
+        mv -v "${PERSISTENT_STORE}/previous" "${PERSISTENT_STORE}/current"
+        comp_start "${PERSISTENT_STORE}/current" 2>&1
+    fi
+}
+
+cleanup() {
+    local rc=0
+
+    if ! test -d "${PERSISTENT_STORE}/cleanup"; then
+        return 0
+    fi
+
+    local image_id
+    while read -r image_id; do
+        if ! grep -qF "$image_id" "${PERSISTENT_STORE}/current/image_ids" 2> /dev/null; then
+            $DOCKER_CMD rmi "$image_id" || rc=1
+        fi
+    done < "${PERSISTENT_STORE}/cleanup/image_ids"
+    rm -rf "${PERSISTENT_STORE}/cleanup"
+
+    return $rc
+}
+
+case "$STATE" in
+    NeedsArtifactReboot)
+        echo "No"
+        ;;
+
+    SupportsRollback)
+        echo "Yes"
+        ;;
+
+    ArtifactInstall)
+        discover_requirements
+        parse_metadata "$FILES"/header/meta-data "$FILES"/header/header-info
```

Comment:
All the below states (inc Cleanup if my comment is correct) do these two steps. Could we move them out of the `case`?

On the one hand having it here explicitly reads better. On the other hand repetition may lead to errors when maintaining it in the future. So up to you!

Reply:
The update module is called with other states as well. I'd rather keep it explicit in these explicitly-supported states. Of course, we can refactor this in the future.